src/node/miner.cpp

```diff
@@ -449,6 +454,104 @@ void BlockAssembler::addPackageTxs(int& nPackagesSelected, int& nDescendantsUpda
     }
 }
 
+class BlockTemplateCacheImpl final: public BlockTemplateCache
+{
+public:
+    BlockTemplateCacheImpl(CTxMemPool* mempool, ChainstateManager& chainman, size_t cache_size);
+    std::pair<std::shared_ptr<CBlockTemplate>, NodeClock::time_point> GetBlockTemplate(const BlockAssembler::Options& options, std::chrono::seconds interval)
+        override EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
+
+    /** Overridden from CValidationInterface. */
+    void BlockConnected(ChainstateRole /*unused*/, const std::shared_ptr<const CBlock>& /*unused*/, const CBlockIndex * /*unused*/)
+        override EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
+private:
+     std::pair<std::shared_ptr<CBlockTemplate>, NodeClock::time_point> CreateBlockTemplateInternal(const BlockAssembler::Options& options) EXCLUSIVE_LOCKS_REQUIRED(m_mutex);
+
+    struct Template {
+        BlockAssembler::Options m_options;
+        NodeClock::time_point m_time_generated;
+        std::shared_ptr<CBlockTemplate> m_block_template;
+
+        bool IntervalElapsed(const std::chrono::seconds& interval) const;
+        bool OptionsEqual(const BlockAssembler::Options& options) const;
+    };
+
+    std::deque<Template> m_block_templates;
+    CTxMemPool* m_mempool{nullptr};
+    ChainstateManager& m_chainman;
+    size_t m_cache_size;
+    mutable Mutex m_mutex;
+};
+
+BlockTemplateCacheImpl::BlockTemplateCacheImpl(CTxMemPool* mempool, ChainstateManager& chainman, size_t cache_size)
+    : m_mempool(mempool), m_chainman(chainman), m_cache_size(cache_size) {}
+
+std::shared_ptr<BlockTemplateCache> MakeBlockTemplateCache(CTxMemPool* mempool, ChainstateManager& chainman, size_t cache_size)
+{
+    return std::make_shared<BlockTemplateCacheImpl>(mempool, chainman, cache_size);
+}
+
+void BlockTemplateCacheImpl::BlockConnected(
+    ChainstateRole /*unused*/,
+    const std::shared_ptr<const CBlock>& /*unused*/,
+    const CBlockIndex* /*unused*/)
+{
+    LOCK(m_mutex);
+    m_block_templates.clear();
+}
+
+bool BlockTemplateCacheImpl::Template::IntervalElapsed(const std::chrono::seconds& interval) const
+{
+    auto now = NodeClock::now();
+    auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - m_time_generated);
+    if (elapsed < std::chrono::seconds::zero()) {
+        // m_time_generated is in the future; node clock is bumped to the past
+        // hence no way to check the elapsed time just assume interval elapsed.
+        return true;
+    }
+    return elapsed >= interval;
+}
+
+bool BlockTemplateCacheImpl::Template::OptionsEqual(const BlockAssembler::Options& other) const
```

Comment:
In commit "node: add block template cache to miner" (4de7887ca47197654f480471db04750d44e405de)

Would seem nice to implemented this as `operator==` helpers or other functions directly attached to BlockCreateOptions and `BlockAssembler::Options` structs. This way if new members are added to the structs the function definitions should not get out of date.

Otherwise there should be comments in the two structs pointing out that this function may need to be updated if new fields are added.

I think I'd also suggest introducing options-comparing functionality in a separate commit to make this commit a little simpler.